Problem Statement- 
Build a decision tree classifier to predict whether a customer will purchase a product or service based on their demographic and behavioral data.
    Use a dataset such as the Bank Marketing dataset from the UCI Machine Learning Repository.

## Importing the neccessary libraries

In [1]:
import pandas as pd 
import matplotlib.pyplot as plt 
%matplotlib inline

In [2]:
df = pd.read_csv("bank-additional-full.csv", sep=';') 
df

age          job  marital            education  default housing loan  \
0       56    housemaid  married             basic.4y       no      no   no   
1       57     services  married          high.school  unknown      no   no   
2       37     services  married          high.school       no     yes   no   
3       40       admin.  married             basic.6y       no      no   no   
4       56     services  married          high.school       no      no  yes   
...    ...          ...      ...                  ...      ...     ...  ...   
41183   73      retired  married  professional.course       no     yes   no   
41184   46  blue-collar  married  professional.course       no      no   no   
41185   56      retired  married    university.degree       no     yes   no   
41186   44   technician  married  professional.course       no      no   no   
41187   74      retired  married  professional.course       no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
0      telephone   may         mon  ...         1    999         0   
1      telephone   may         mon  ...         1    999         0   
2      telephone   may         mon  ...         1    999         0   
3      telephone   may         mon  ...         1    999         0   
4      telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
41183   cellular   nov         fri  ...         1    999         0   
41184   cellular   nov         fri  ...         1    999         0   
41185   cellular   nov         fri  ...         2    999         0   
41186   cellular   nov         fri  ...         1    999         0   
41187   cellular   nov         fri  ...         3    999         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent          1.1          93.994          -36.4      4.857   
1      nonexistent          1.1          93.994          -36.4      4.857   
2      nonexistent          1.1          93.994          -36.4      4.857   
3      nonexistent          1.1          93.994          -36.4      4.857   
4      nonexistent          1.1          93.994          -36.4      4.857   
...            ...          ...             ...            ...        ...   
41183  nonexistent         -1.1          94.767          -50.8      1.028   
41184  nonexistent         -1.1          94.767          -50.8      1.028   
41185  nonexistent         -1.1          94.767          -50.8      1.028   
41186  nonexistent         -1.1          94.767          -50.8      1.028   
41187      failure         -1.1          94.767          -50.8      1.028   

       nr.employed    y  
0           5191.0   no  
1           5191.0   no  
2           5191.0   no  
3           5191.0   no  
4           5191.0   no  
...            ...  ...  
41183       4963.6  yes  
41184       4963.6   no  
41185       4963.6   no  
41186       4963.6  yes  
41187       4963.6   no  

[41188 rows x 21 columns]

In [3]:
df_new = df.drop(["campaign","pdays","previous","poutcome","emp.var.rate","cons.price.idx","cons.conf.idx","euribor3m","nr.employed","duration","contact","default"],axis=1)

In [4]:
df_new.head()

age        job  marital    education housing loan month day_of_week   y
0   56  housemaid  married     basic.4y      no   no   may         mon  no
1   57   services  married  high.school      no   no   may         mon  no
2   37   services  married  high.school     yes   no   may         mon  no
3   40     admin.  married     basic.6y      no   no   may         mon  no
4   56   services  married  high.school      no  yes   may         mon  no

In [5]:
df_new.describe

<bound method NDFrame.describe of        age          job  marital            education housing loan month  \
0       56    housemaid  married             basic.4y      no   no   may   
1       57     services  married          high.school      no   no   may   
2       37     services  married          high.school     yes   no   may   
3       40       admin.  married             basic.6y      no   no   may   
4       56     services  married          high.school      no  yes   may   
...    ...          ...      ...                  ...     ...  ...   ...   
41183   73      retired  married  professional.course     yes   no   nov   
41184   46  blue-collar  married  professional.course      no   no   nov   
41185   56      retired  married    university.degree     yes   no   nov   
41186   44   technician  married  professional.course      no   no   nov   
41187   74      retired  married  professional.course     yes   no   nov   

      day_of_week    y  
0             mon   no  
1  

In [6]:
inputs = df_new.drop(["y"],axis=1)
target= df.y

In [7]:
# As the ml model better works on the numerical data we have to convert our categorical varibles into the numrical 
# we do this by using the lable encoder and one hot encoding 
from sklearn.preprocessing import LabelEncoder
le_marital = LabelEncoder()
le_job = LabelEncoder()
le_education = LabelEncoder()
le_housing = LabelEncoder()
le_loan = LabelEncoder()
le_month = LabelEncoder()
le_day_of_week = LabelEncoder()

In [8]:
inputs['marital_n'] = le_marital.fit_transform(inputs['marital'])
inputs['job_n'] = le_job.fit_transform(inputs['job'])
inputs['education_n'] = le_education.fit_transform(inputs['education'])
inputs['housing_n'] = le_housing.fit_transform(inputs['housing'])
inputs['loan_n'] = le_loan.fit_transform(inputs['loan'])

In [9]:
inputs

age          job  marital            education housing loan month  \
0       56    housemaid  married             basic.4y      no   no   may   
1       57     services  married          high.school      no   no   may   
2       37     services  married          high.school     yes   no   may   
3       40       admin.  married             basic.6y      no   no   may   
4       56     services  married          high.school      no  yes   may   
...    ...          ...      ...                  ...     ...  ...   ...   
41183   73      retired  married  professional.course     yes   no   nov   
41184   46  blue-collar  married  professional.course      no   no   nov   
41185   56      retired  married    university.degree     yes   no   nov   
41186   44   technician  married  professional.course      no   no   nov   
41187   74      retired  married  professional.course     yes   no   nov   

      day_of_week  marital_n  job_n  education_n  housing_n  loan_n  
0             mon          1      3            0          0       0  
1             mon          1      7            3          0       0  
2             mon          1      7            3          2       0  
3             mon          1      0            1          0       0  
4             mon          1      7            3          0       2  
...           ...        ...    ...          ...        ...     ...  
41183         fri          1      5            5          2       0  
41184         fri          1      1            5          0       0  
41185         fri          1      5            6          2       0  
41186         fri          1      9            5          0       0  
41187         fri          1      5            5          2       0  

[41188 rows x 13 columns]

In [10]:
inputs_n = inputs.drop(["job","marital","education","housing","loan","month","day_of_week"],axis=1)
inputs_n

age  marital_n  job_n  education_n  housing_n  loan_n
0       56          1      3            0          0       0
1       57          1      7            3          0       0
2       37          1      7            3          2       0
3       40          1      0            1          0       0
4       56          1      7            3          0       2
...    ...        ...    ...          ...        ...     ...
41183   73          1      5            5          2       0
41184   46          1      1            5          0       0
41185   56          1      5            6          2       0
41186   44          1      9            5          0       0
41187   74          1      5            5          2       0

[41188 rows x 6 columns]

In [11]:

from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(inputs_n,target,test_size=0.2,random_state=42)

In [12]:
len(x_test)

8238

In [13]:
len(x_train)

32950

In [14]:
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier(criterion='entropy',random_state=0)


In [15]:
clf.fit(x_train,y_train)

DecisionTreeClassifier(criterion='entropy', random_state=0)

In [16]:
clf.score(x_test,y_test)

0.8692643845593591

In [17]:
clf = DecisionTreeClassifier(criterion='gini',random_state=0)

In [18]:
clf.fit(x_train,y_train)

DecisionTreeClassifier(random_state=0)

In [19]:
clf.score(x_test,y_test)

0.8676863316338917

In [20]:
y_pred = clf.predict(x_test)

In [21]:
from sklearn.model_selection import cross_val_score
cross_val_score(clf,inputs_n,target, cv=10)

array([0.87424132, 0.86890022, 0.86792911, 0.869143  , 0.85554746,
       0.86817189, 0.86962855, 0.85166303, 0.87323944, 0.76177756])

In [22]:
from sklearn.metrics import confusion_matrix, classification_report

In [23]:
conf_matrix = confusion_matrix(y_test, y_pred)

In [24]:
print("Confusion Matrix:")
print(conf_matrix)

#classification report (precision, recall, F1-score, support)
class_report = classification_report(y_test, y_pred)
print("\nClassification Report:")
print(class_report)

Confusion Matrix:
[[7074  229]
 [ 861   74]]

Classification Report:
              precision    recall  f1-score   support

          no       0.89      0.97      0.93      7303
         yes       0.24      0.08      0.12       935

    accuracy                           0.87      8238
   macro avg       0.57      0.52      0.52      8238
weighted avg       0.82      0.87      0.84      8238



## Conclusion 

The overall accuracy is 0.87 or 87%, which means that the model correctly predicted 87% of the instances in the test set.
    
As our model well to predict the no class with a high precision of 89% 

The models performance for the "yes" class is relatively poor, with a low precision of 24%